lib/open_feature/client.rb

```diff
@@ -218,6 +235,28 @@ def fetch_structure_details(flag_key:, default_value:, context: nil, options: ni
     sig { returns(Provider) }
     attr_reader :provider
 
+    sig do
+      type_parameters(:U).params(
+        flag_key: String,
+        default_value: T.type_parameter(:U),
+        invocation_context: T.nilable(EvaluationContext),
+        options: T.nilable(EvaluationOptions),
+        flag_type: String
+      ).returns(EvaluationContext)
+    end
+    def build_context_with_before_hooks(flag_key:, default_value:, invocation_context:, options:, flag_type:)
+      evaluation_context = build_context(invocation_context) || OpenFeature::EvaluationContext.new
+      hook_context = HookContext.new(flag_key: flag_key, default_value: default_value,
+                                     evaluation_context: evaluation_context, flag_type: flag_type)
+      OpenFeature::Hook::BeforeHook.call(hooks: sequence_before_hooks(options), context: hook_context,
```

Comment:
I would hope this would read a little more like `sequence_hooks.reduce(evaluation_context) { |eval_context, hook| eval_context.merge(hook.call(hook_context: hook_context)) }

Reply:
I think I agree that doing this in-line here is a bit cleaner.

Reply:
hmm the only sticky point is that the test suite is based on the helper class methods currently. I don't think they should be public methods on the `Client`. 

thoughts?

Reply:
We'll want to pull some of these private methods into a command type pattern or add them as functionality on the `Hook` module. https://github.com/maxveldink/openfeature-sdk-sorbet/blob/main/lib/open_feature/evaluation_context_builder.rb is an example where I did something similar for ordering EvaluationContext merging

Reply:
yes that makes sense! will take a look at this next week

Reply:
I took a slightly different approach but same idea I think and created a `Hooks` struct. LMK your thoughts.